### ISSUE TITLE
ANN: Quill--write/play Ink files in a browser

### ISSUE THREAD

**Issue by MattConrad** [inkle/ink#74]
I've been playing around with Ink and the new ASP.NET Core and have an application up that runs Ink stories in the browser:

site: http://jeejah.xyz/quill/
source: https://github.com/MattConrad/Quill

Quill uses the regular .NET/C# runtime, but with a browser as the front end. (It's not the javascript port you're looking for.)

It seems to work pretty well (hasn't crashed accidentally yet). I've noticed one quirk with the Whack A Mole example--the options do not progressively disappear the way they would if you were running in the console. Quill reloads both the story and the saved game state on every request, and I imagine this is why, but I haven't dug into it to check. Might be some other oddities.

You can't see changes to an .ink immediately, but there is a Replay Choices option so you can make edits and quickly reload the new revised story up to where you left off.

**Comment by joethephish** [comment]
That's super awesome, thanks for sharing!

Any chance you could get it to echo the compile errors/warnings that are produced when making syntax errors?

Also, I find that sometimes the layout doesn't work entirely perfectly. I first tried it on mobile, and some of the elements were really spaced out (reproducible by making your browser window narrower). Also, the choice links seem to be squashed against the right hand side, when I'd suggest they should come underneath the story content?

But great work in general, it's exciting to have a scratch pad for people to play with ink!

**Comment by MattConrad** [comment]
Thanks for the suggestions, these sound like good ideas. I'll see about making some revisions.

**Comment by MattConrad** [comment]
Some changes applied, pretty much what you suggested. Ink syntax errors are now caught and displayed to the user. Story choices appear under the story text rather than alongside, and story width is widened. Messed up layout is improved on mobile. A few other little tweaks. Added micabytes's Cloak Of Darkness to the examples.

Joe, can you try it out on mobile and see if the flow makes better sense? I'm a little concerned that people will click Play on mobile and not realize the story text is appearing under the Play button. I can put a scroll-into-view for this, but that is just annoying if people do understand what's happening without it. (I basically don't browse on mobile, so don't have a good sense of user expectations.)

**Comment by joethephish** [comment]
Thanks, the changes work great! On mobile it's much better. It's slightly fiddly because you need to scroll down within the code view to get to the play area, but not sure whether there's much that can be done about that without a more radical redesign.

Awesome being able to get warnings/errors on the page now. You might want to wrap the error lines in `<p>...</p>` tags though, since they're all concatenating on the same line right now.

Also, I was wondering whether it might be worth dropping [Ace](https://ace.c9.io/) in so that there's a nicer code editing area built in, complete with line numbers? I haven't used Ace before, but it might even be possible to get it to highlight the line where the error is...

Ooh, and I don't know how you feel about this from the point of view of the server, but it would be incredible if it auto-compiled after edits? :-)

Obviously I can feature-creep all day if you're interested and enjoying the project! :-D

**Comment by MattConrad** [comment]
Thanks for the further suggestions, I'll look into these. I did look at some line numbering textareas, but none of them worked well. Ace isn't one that I tried, though. I'll check it out.

When you say auto complied after edits, you mean a certain time after the user stops typing the story is automatically auto-submitted and we attempt to replay the previous choices, yes? Not sure I'll try to do that one, at least not soon, but want to make sure I understand the idea correctly.

**Comment by joethephish** [comment]
In JavaScript you can get events for when a text field gets changes or loses focus, and from that you could fire off your compile event. You could also limit it so that it's only if the user edits and then pauses for a set interval, or make it so that it queues up until it's had a response from a previous compile request. But yeah, I totally understand that there are potentially inherent problems with it, and it's a bit more complex!

**Comment by MattConrad** [comment]
More Quill improvements applied. 

Ace did indeed work as an Ink editor. You now get line numbers and, if you have errors, annotations of the errors on the failing lines. This is pretty neat and I'm happy you pointed me in the direction of Ace.

You can now check a box to have your Ink changes automatically recompiled as you're going along. If you have autoplay checked, making a change and then pausing for a couple seconds will trigger an automatic recompile.

Another user asked for a way to get a permanent link that other people to use to play a story without having to see/interact with the source code. Get Permalink becomes available after the first successful compile.

If you bump into issues, let me know at https://github.com/MattConrad/Quill/issues .

**Comment by joningold** [comment]
This is awesome. I've been using Quill quite a bit now ( mostly admittedly
to test minimal sources while bug hunting, but still. ) fantastic resource.

Jon

**Comment by MattConrad** [comment]
Oh, nice. Yes, bringing InkMode to Quill will be very good. Thanks for the notice. On the list.

Congrats on Inky . . . and I also see inkjs is out there now. Everything's coming up Inkle!

**Comment by Glidias** [comment]
Have you thought of doing up a compiler/player service via NodeJS and EdgeJS? NodeJS child_process to compile files via the inklecate.exe (inklecate.exe  execute for windows or "mono" it directly for linux/ubuntu servers), and  EdgeJS to run the dll library and link it to Javascript/Node for runtime.

You can preview my wip inkle combat script here...
http://ec2-54-254-243-13.ap-southeast-1.compute.amazonaws.com:3000/
and the sources at https://github.com/Glidias/trosink/tree/master/edge  . My plan is to have a scenerio creator to procedurally generate inkle files, compile them, and play scenerio.

**Comment by y-lohse** [comment]
FWIW, you could use [inkjs](https://github.com/y-lohse/inkjs) for the "playing the story" part. It works in node and in the browser, so you can decide where you want to "play" the story.

I'm thinking of shipping the inklecate binary with the node package, just because it would be a convenient way to have a system-wide installation.

**Comment by Glidias** [comment]
yup,  inkjs would be the ideal cost-effective solution for JS... being able to run offline(anywhere) as well with fewer dependencies. (even better if the compiler could also be ported over as well rather than require node to compile it.) ... But hmm, once it can get edge/scripts/stories/tros.ink.json to run though.

**Comment by y-lohse** [comment]
It should, although apparently it doesn't. The ink files are a bit too massive for me to make sense of, but if  you can provide smaller test cases of things that are not working, I'd be happy to fix them!

**Comment by lhughes41** [comment]
Matt,
 First thanks for a really useful tool which I and my team having been making great use of in easily getting feedback on Ink writing.  That being said, we are most distressed to find it down this morning. Right as we were hoping to use it again.  May we hope it will be restored soon? 
 Thanks again,
  Luke

**Comment by MattConrad** [comment]
Luke, thanks for pointing out the error.

Should be fixed now. 

At least until the thing that shouldn't have happened in the first place happens again. Maybe it was a fluke. Please do remark if you get similar errors again.

**Comment by lhughes41** [comment]
Working!  Thanks for the fix!

**Comment by tameTNT** [comment]
This is amazing as some one who just wants to write some ideas up as they come come to them quickly on their phone in their browser! Keep up the great development work!
